File: cass/db/queries.py
"""Enriched queries and shared query dataset helpers."""

from __future__ import annotations

__docformat__ = "google"

import sqlite_utils

from .core import _SAFE_IDENT_RE, get_db

# ---------------------------------------------------------------------------
# Enriched queries — JOINed views used by the viewer and tests
# ---------------------------------------------------------------------------

ENRICHED_QUERIES: dict[str, str] = {
    "canvas_submissions": """
        SELECT
            cs.canvas_user_id,
            cs.canvas_assignment_id,
            st.sortable_name AS student,
            ca.name AS assignment_name,
            ca.assignment_group,
            CASE WHEN cs.submitted_at IS NOT NULL
                THEN cs.submitted_at
                ELSE '' END AS submitted_at,
            ca.due_at,
            cs.score,
            cs.workflow_state,
            cs.late
        FROM canvas_submissions cs
        LEFT JOIN canvas_students st ON cs.canvas_user_id = st.canvas_id
        LEFT JOIN canvas_assignments ca
            ON cs.canvas_assignment_id = ca.canvas_id
    """,
    "canvas_students": """
        SELECT
            cs.sortable_name AS student,
            cs.email,
            cs.canvas_id
        FROM canvas_students cs
        ORDER BY cs.sortable_name
    """,
    "canvas_assignments": """
        SELECT
            ca.assignment_group,
            ca.name,
            ca.points_possible,
            ca.due_at,
            ca.published,
            ca.canvas_id
        FROM canvas_assignments ca
        ORDER BY ca.assignment_group, ca.name
    """,
    "canvas_grades": """
        SELECT
            cg.canvas_user_id,
            cg.canvas_assignment_id,
            st.sortable_name AS student,
            ca.name AS assignment_name,
            ca.assignment_group,
            cg.score,
            cg.posted_grade,
            cg.updated_at
        FROM canvas_grades cg
        LEFT JOIN canvas_students st ON cg.canvas_user_id = st.canvas_id
        LEFT JOIN canvas_assignments ca
            ON cg.canvas_assignment_id = ca.canvas_id
    """,
}

QUERY_DATASETS = ("students", "assignments", "submissions", "gradebook")
TABLE_QUERY_DATASETS = {
    "students": "canvas_students",
    "assignments": "canvas_assignments",
}
SUBMISSIONS_BASE_QUERY = """
    SELECT * FROM (
      SELECT
        st.sortable_name AS student,
        ca.name AS assignment,
        cs.submitted,
        cs.submitted_at,
        cs.late,
        cs.lateness_seconds,
        cs.score,
        cs.workflow_state
      FROM canvas_submissions cs
      JOIN canvas_students st ON st.canvas_id = cs.canvas_user_id
      JOIN canvas_assignments ca ON ca.canvas_id = cs.canvas_assignment_id
    ) submissions
"""


def build_submissions_query(
    *,
    where: str = "",
    order: str = "",
    limit: int = 0,
) -> str:
    """Build the shared submissions dataset query."""
    query = SUBMISSIONS_BASE_QUERY
    if where:
        query += f" WHERE {where}"
    query += f" ORDER BY {order or 'assignment, student'}"
    if limit:
        query += f" LIMIT {limit}"
    return query


def get_enriched_rows(
    sdb: sqlite_utils.Database,
    table: str,
) -> list[dict[str, object]]:
    """Return all rows from a table using enriched query if available."""
    if not _SAFE_IDENT_RE.match(table):
        msg = f"Invalid table name: {table!r}"
        raise ValueError(msg)
    query = ENRICHED_QUERIES.get(table, f"SELECT * FROM {table}")
    result = sdb.execute(query)
    col_names = [desc[0] for desc in result.description]
    rows = result.fetchall()
    return [dict(zip(col_names, row, strict=True)) for row in rows]


# ---------------------------------------------------------------------------
# Raw query (for `cass query`)
# ---------------------------------------------------------------------------


class QueryResult:
    """Thin wrapper around a sqlite3 cursor for Rich table rendering."""

    def __init__(self, columns: list[str], rows: list[tuple[object, ...]]) -> None:
        self.columns = columns
        self._rows = rows

    def fetchall(self) -> list[tuple[object, ...]]:
        return self._rows


def sql(query: str) -> QueryResult:
    """Execute SQL and return a QueryResult with .columns and .fetchall()."""
    sdb = get_db()
    cur = sdb.execute(query)
    cols = [desc[0] for desc in cur.description] if cur.description else []
    return QueryResult(cols, cur.fetchall())


def run_query(query: str) -> list[dict[str, object]]:
    """Execute arbitrary SQL and return results as list of dicts."""
    sdb = get_db()
    return list(sdb.query(query))

File: src/db/queries.ts
/**
 * Typed queries — replaces manual SQL builders from Python queries.py.
 * Uses kysely JOINs for enriched datasets.
 */
import { type Kysely, type SqlBool, sql } from "kysely";
import { match } from "ts-pattern";
import type { Database } from "./schema.ts";

export type Dataset = "students" | "assignments" | "submissions" | "gradebook";

export interface QueryOptions {
	where?: string;
	order?: string;
	limit?: number;
}

/** Enriched students: master students with all available fields. */
export async function queryStudents(db: Kysely<Database>, opts: QueryOptions = {}) {
	let query = db.selectFrom("students").selectAll("students");

	if (opts.where) query = query.where(sql.raw<SqlBool>(opts.where));
	if (opts.order) query = query.orderBy(sql.raw(opts.order));
	if (opts.limit) query = query.limit(opts.limit);

	return query.execute();
}

/** Enriched assignments: master assignments with all available fields. */
export async function queryAssignments(db: Kysely<Database>, opts: QueryOptions = {}) {
	let query = db.selectFrom("assignments").selectAll("assignments");

	if (opts.where) query = query.where(sql.raw<SqlBool>(opts.where));
	if (opts.order) query = query.orderBy(sql.raw(opts.order));
	if (opts.limit) query = query.limit(opts.limit);

	return query.execute();
}

/** Submissions joined with student names and assignment titles. */
export async function querySubmissions(db: Kysely<Database>, opts: QueryOptions = {}) {
	let query = db
		.selectFrom("canvas_submissions as cs")
		.innerJoin("students as s", "s.canvas_id", "cs.canvas_user_id")
		.innerJoin("assignments as a", "a.canvas_assignment_id", "cs.canvas_assignment_id")
		.select([
			"s.name",
			"a.title",
			"a.slug",
			"cs.canvas_user_id",
			"cs.canvas_assignment_id",
			"cs.submitted",
			"cs.score",
			"cs.late",
			"cs.workflow_state",
			"cs.submitted_at",
		]);

	if (opts.where) query = query.where(sql.raw<SqlBool>(opts.where));
	if (opts.order) query = query.orderBy(sql.raw(opts.order));
	if (opts.limit) query = query.limit(opts.limit);

	return query.execute();
}

/** Dispatch to the appropriate dataset query using ts-pattern. */
export function queryDataset(
	db: Kysely<Database>,
	dataset: Dataset,
	opts: QueryOptions = {},
): Promise<Record<string, unknown>[]> {
	return match(dataset)
		.with("students", () => queryStudents(db, opts))
		.with("assignments", () => queryAssignments(db, opts))
		.with("submissions", () => querySubmissions(db, opts))
		.with("gradebook", () => queryGradebook(db, opts))
		.exhaustive() as Promise<Record<string, unknown>[]>;
}

/** Gradebook — placeholder for the matrix pivot view. */
export async function queryGradebook(db: Kysely<Database>, opts: QueryOptions = {}) {
	let query = db
		.selectFrom("canvas_submissions as cs")
		.innerJoin("students as s", "s.canvas_id", "cs.canvas_user_id")
		.innerJoin("assignments as a", "a.canvas_assignment_id", "cs.canvas_assignment_id")
		.select(["s.name", "a.title", "cs.score"]);

	if (opts.where) query = query.where(sql.raw<SqlBool>(opts.where));
	if (opts.order) query = query.orderBy(sql.raw(opts.order));
	if (opts.limit) query = query.limit(opts.limit);

	return query.execute();
}
